core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java

```diff
@@ -1232,8 +1203,218 @@ public void run() {
+            //operations, multiple prompts from RS is received. Hence check to see if there
+            //is a TS object in main TS list and transaction is still active.
+            //Note that tsEntry is local TS object. 
+            if (hbtx.mapTransactionStates.get(txID) != null) {
+              if (hbtx.mapTransactionStates.get(txID).getStatus().toString().contains("ACTIVE")) {
+                isTransactionStillAlive = 1;
+              }
+              if (LOG.isInfoEnabled()) 
+              LOG.info("TRAF RCOV THREAD: TID " + txID
+                        + " still has TS object in TM memory. TS details: "
+                        + hbtx.mapTransactionStates.get(txID).toString() 
+                        + " transactionAlive: " + isTransactionStillAlive);
+              if(isTransactionStillAlive == 1)
+                continue; //for loop
+            }
+           
+            try {
+                audit.getTransactionState(ts);
+                if (ts.getStatus().equals(TransState.STATE_COMMITTED.toString())) {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("TRAF RCOV THREAD:Redriving commit for " + txID + " number of regions " + ts.getParticipatingRegions().size() +
+                                " and tolerating UnknownTransactionExceptions");
+                    txnManager.doCommit(ts, true /*ignore UnknownTransactionException*/);
+                    if(useTlog && useForgotten) {
+                        long nextAsn = tLog.getNextAuditSeqNum((int)TransactionState.getNodeId(txID));
+                        tLog.putSingleRecord(txID, ts.getCommitId(), "FORGOTTEN", null, forceForgotten, nextAsn);
+                    }
+                } else if (ts.getStatus().equals(TransState.STATE_ABORTED.toString())) {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("TRAF RCOV THREAD:Redriving abort for " + txID);
+                    txnManager.abort(ts);
+                } else {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("TRAF RCOV THREAD:Redriving abort for " + txID);
+                    LOG.warn("Recovering transaction " + txID + ", status is not set to COMMITTED or ABORTED. Aborting.");
+                    txnManager.abort(ts);
+                }
+  
+            } catch (UnsuccessfulDDLException ddle) {
+                LOG.error("UnsuccessfulDDLException encountered by Recovery Thread. Registering for retry. txID: " + txID + "Exception " , ddle);
+  
+                //Note that there may not be anymore redrive triggers from region server point of view for DDL operation.
+                //Register this DDL transaction for subsequent redrive from Audit Control Event.
+                //TODO: Launch a new Redrive Thread out of auditControlPoint().
+                TmDDL tmDDL = hbtx.getTmDDL();
+                tmDDL.setState(txID,"REDRIVE");
```

Comment:
@shangshengtung :  There needs to be clear distinction of responsibility between worker thread and recovery thread. Here there is not need to change the state of txID in tmDDL. Recovery thread can retry in subsequent loop.